# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1497645527 14400
#      Fri Jun 16 16:38:47 2017 -0400
# Node ID fca2f775ab4cc223d2e2cd7e63e58832d87ade9d
# Parent  0a6e87804cabfbd80b903b952f1da84b98603067
eliminate memory leak when parsing local functions in a classdef file

* oct-parse.in.yy (file): Delete dummy statement list created from
parsing local functions in a classdef file.

diff --git a/libinterp/parse-tree/oct-parse.in.yy b/libinterp/parse-tree/oct-parse.in.yy
--- a/libinterp/parse-tree/oct-parse.in.yy
+++ b/libinterp/parse-tree/oct-parse.in.yy
@@ -1448,17 +1448,24 @@ file            : INPUT_FILE opt_nl opt_
                       }
 
                     $$ = 0;
                   }
                 | INPUT_FILE opt_nl classdef parsing_local_fcns opt_sep opt_fcn_list END_OF_INPUT
                   {
                     YYUSE ($2);
                     YYUSE ($5);
-                    YYUSE ($6);
+
+                    // Delete the dummy statement_list we created
+                    // after parsing the function.  Any local function
+                    // definitions found in the file have already been
+                    // stored in the symbol table or in
+                    // base_parser::primary_fcn_ptr.
+
+                    delete $6;
 
                     if (lexer.reading_classdef_file)
                       parser.classdef_object = $3;
 
                     $$ = 0;
                   }
                 ;
 
